### PR TITLE
introduce savepoint

### PR DESCRIPTION
**Pull request by BusyJay**
@siddontang @zhangjinpeng1987 @hhkbp2 PTAL

### CODE REVIEW

**Comment by ngaut** [comment]
How we are going to use this feature ?

**Comment by BusyJay** [comment]
When doing batch apply, there may be errors in the middle, so it has to rollback to maintain atomicity.

File: src/rocksdb.rs [hunk 1]
```diff
@@ -1301,6 +1314,22 @@ mod test {
...
+        // test save point
+        let mut batch = WriteBatch::new();
+        batch.put(b"k10", b"v10").unwrap();
+        batch.set_save_point();
+        batch.put(b"k11", b"v11").unwrap();
+        batch.put(b"k12", b"v12").unwrap();
+        batch.rollback_to_save_point().unwrap();
+        let p = db.write(batch);
+        assert!(p.is_ok());
+        let r: Result<Option<DBVector>, String> = db.get(b"k10");
```
Comment (siddontang) [inline src/rocksdb.rs]: need to declare the type here?

File: src/rocksdb.rs [hunk 2]
```diff
@@ -1301,6 +1314,22 @@ mod test {
...
+        assert!(r.unwrap().unwrap().to_utf8().unwrap() == "v10");
```
Comment (siddontang) [inline src/rocksdb.rs]: seem that we can use:
```
assert!(r.unwrap().unwrap(), b"v10");
```

File: src/rocksdb.rs [hunk 3]
```diff
@@ -1301,6 +1314,22 @@ mod test {
         assert!(batch.data_size() > prev_size);
         batch.clear();
         assert_eq!(batch.data_size(), prev_size);
+
+        // test save point
+        let mut batch = WriteBatch::new();
+        batch.put(b"k10", b"v10").unwrap();
+        batch.set_save_point();
```
Comment (zhangjinpeng1987) [inline src/rocksdb.rs]: If I set save point multi times, if the new save point will override the old one?

Comment (zhangjinpeng1987) [inline src/rocksdb.rs]: Please add test for set multi save points.

**Comment by zhangjinpeng1987** [comment]
rest LGTM

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**